### PR TITLE
bpo-36611: Disable serialno field of debug memory allocators

### PR DESCRIPTION
**Pull request by vstinner**
Omit serialno field from debug hooks on Python memory allocators to
reduce the memory footprint by 5%.

Enable tracemalloc to get the traceback where a memory block has been
allocated when a fatal memory error is logged to decide where to put
a breakpoint.

Compile Python with PYMEM_DEBUG_SERIALNO defined to get back the
field.

https://bugs.python.org/issue36611

### CODE REVIEW

**Review by methane — APPROVED** [review]

**Comment by vstinner** [comment]
@methane: Just to be sure, you prefer this approach (disable by default, but keep the code) over PR #12795 which simply removes the code?

I have also a preference for this approach: keep the code.

File: Objects/obmalloc.c [hunk 1]
```diff
@@ -2110,47 +2130,60 @@ _PyMem_DebugRawRealloc(void *ctx, void *p, size_t nbytes)
...
+               original_nbytes + PYMEM_DEBUG_EXTRA_BYTES);
     }
     else {
         memcpy(save, data, ERASED_SIZE);
-        memset(head, DEADBYTE, ERASED_SIZE + 2*SST);
+        memset(head, DEADBYTE, ERASED_SIZE + 2 * SST);
         memcpy(&save[ERASED_SIZE], tail - ERASED_SIZE, ERASED_SIZE);
-        memset(tail - ERASED_SIZE, DEADBYTE, ERASED_SIZE + 2*SST);
+#ifdef PYMEM_DEBUG_SERIALNO
+        memset(tail - ERASED_SIZE, DEADBYTE, ERASED_SIZE + 2 * SST);
```
Comment (serhiy-storchaka) [inline Objects/obmalloc.c]: Or
```
memset(tail - ERASED_SIZE, DEADBYTE,
       ERASED_SIZE + PYMEM_DEBUG_EXTRA_BYTES - 2 * SST);
```

**Review by serhiy-storchaka — APPROVED** [review]
LGTM. Did you test with PYMEM_DEBUG_SERIALNO defined?

**Comment by serhiy-storchaka** [comment]
> Just to be sure, you prefer this approach (disable by default, but keep the code) over PR #12795 which simply removes the code?

It is up to you as the author and the main user of this code.

**Comment by vstinner** [comment]
Oh, Travis CI failed with a crash in test_PyThreadState_SetAsyncExc()... I'm not sure if it's related to my change?

```
test_PyThreadState_SetAsyncExc (test.test_threading.ThreadTests) ...     started worker thread
    trying nonsensical thread id
Fatal Python error: Segmentation fault
Thread 0x00007f7df4ff9700 (most recent call first):
  File "/home/travis/build/python/cpython/Lib/test/test_threading.py", line 230 in run
  File "/home/travis/build/python/cpython/Lib/threading.py", line 917 in _bootstrap_inner
  File "/home/travis/build/python/cpython/Lib/threading.py", line 885 in _bootstrap
Current thread 0x00007f7e290da700 (most recent call first):
  File "/home/travis/build/python/cpython/Lib/test/test_threading.py", line 244 in test_PyThreadState_SetAsyncExc
  File "/home/travis/build/python/cpython/Lib/unittest/case.py", line 680 in run
```

https://travis-ci.org/python/cpython/jobs/519191613

File: Objects/obmalloc.c (hunk 1 above)
Comment (vstinner) [reply Objects/obmalloc.c]: Nice trick, I applied it ;-)

**Comment by vstinner** [comment]
> LGTM. Did you test with PYMEM_DEBUG_SERIALNO defined?

I failed to find a smart way to fix test_capi.test_api_misuse() depending if PYMEM_DEBUG_SERIALNO is defined or not, since this define might be only defined in obmalloc.c (not exported)... But I just found one! Since the test uses a regular expression, I just made the line optional in the expected test!

So yes, I just tested and, with my new commit, test_capi pass.

**Comment by zooba** [comment]
Wait, I didn't know about this! Can I use it to debug the dump refs crash? (Where a tuple element is being deallocated even though there's a strong reference to it)

**Comment by methane** [comment]
> @methane: Just to be sure, you prefer this approach (disable by default, but keep the code) over PR #12795 which simply removes the code?

Yes.  And if there are no feedback about it, we can remove it in the future.

**Comment by vstinner** [comment]
Oh wow, that's scary: now it's test_array_from_size() which crashed on Travis CI!

https://travis-ci.org/python/cpython/jobs/519311289

**Comment by vstinner** [comment]
@zooba: "Wait, I didn't know about this! Can I use it to debug the dump refs crash? (Where a tuple element is being deallocated even though there's a strong reference to it)"

I don't understand your question. What do you want to do? This PR is about removing a "serialno" stored in every memory blocks by the debug hooks installed on Python memory allocators. It's lower level than Python objects.

What is the current behavior on your "refs crash"? Which function crash? How?

**Comment by vstinner** [comment]
**Maybe** I got https://bugs.python.org/issue33608 regression by mistake. Just in case, I rebased my PR on top of master to ensure that the regression is reverted.

**Comment by vstinner** [comment]
I tried to reproduce the Travis CI:

```
$ ./configure --with-pydebug CFLAGS=-O3
$ make # use GCC, whereas Travis CI uses clang
```

I ran `` ./python -m test -u all,-gui -j0 -r`` twice in parallel locally: I don't get any issue!? AppVeyor didn't crash. These 2 crashes on Travis CI are really surprising.

**Comment by vstinner** [comment]
> I ran ./python -m test -u all,-gui -j0 -r twice in parallel locally: I don't get any issue!? AppVeyor didn't crash. These 2 crashes on Travis CI are really surprising.

I succeeded to reproduce the crash, but only using clang with -O3. In fact, it's a memory alignement issues. clang makes more assumptions than GCC and Visual Studio. The bug isn't caused by my change, but my change "triggered the bug".

See https://bugs.python.org/issue36618

**Comment by vstinner** [comment]
I pushed my fix for clang: PR #12809. I rebased this PR on top of it.

**Comment by vstinner** [comment]
Thanks for the review @serhiy-storchaka and @methane!

**Pull request merged**